ocaml/lambda/matching.ml

```diff
@@ -129,6 +129,43 @@ let may_compat = MayCompat.compat
 
 and may_compats = MayCompat.compats
 
+type action =
+  | Guarded of
+      { patch_guarded: patch:lambda -> lambda
+      ; unpatched: lambda }
+  (* Guarded actions must allow for fallthrough if the guard fails.
+
+     When translating a guarded action, the code to execute on fallthrough must
+     be "patched" in. Because the fallthrough code is translated after the
+     action is created, guarded actions carry a function [patch_guarded], which
+     generates a lambda term for the action from the fallthrough code.
+
+     Some translation functionality requires us to check syntactic properties
+     of actions. Rather than recomputing [patch_guarded] at the time of these
+     checks, we keep track of [unpatched], a lambda term which contains
+     [staticfail] in the position to be patched.
+  *)
+  | Unguarded of lambda
+
+let mk_guarded ~patch_guarded : action =
```

Comment:
Let's CR-soon the flaw we identified about running `transl_match` twice (it's expensive-ish!).

Reply:
Done!